Add Solidus 4 support

kjriga opened the pull request:
This adds Solidus 4 support while maintaining backwards compatibility.

- Updating the solidus core dependency
- Fixing email subscriber for Solidus 4s refactored mailer system (ensuring compatibility with non-solidus-4)

Specs are passing locally. 
```
All examples were filtered out; ignoring {focus: true}

Randomized with seed 26676
Finished in 1.61 seconds (files took 2.3 seconds to load)
99 examples, 0 failures

Randomized with seed 26676

Coverage report generated for RSpec to /Users/kendrariga/Documents/work-projects/solidus_tracking/coverage.
Line Coverage: 87.58% (402 / 459)
```

Commit:
Bump supported Solidus versions

We want to support Solidus 4 to 4.5.

I stopped at 4.5 because there are changes in 4.6+ with some
'disable_builtin_emails' configuration that required more work, so I
think a plan to move forward is to make the current gemspec only allow
'>= 4.0.0', '< 4.6', and add some deprecation warnings about that
feature becoming obsolete. Then have a new version that's '>= 4.6.0',
'<5.0' that removes it all.

I am doing this work specifically to bump another app that uses it to
Solidus 4, and that is my priority at the moment.

Commit:
Update mailer subscriber for Solidus 4 compatibility

- Update method name from order_finalized to send_confirmation_email
- Change target class from Spree::MailerSubscriber to Spree::OrderMailerSubscriber
- Remove version conditional since gem now only supports Solidus 4+

kjriga commented on spec/models/spree/user_spec.rb:
```diff
@@ -12,7 +12,13 @@
 
+
   describe '#send_reset_password_instructions' do
+    let(:email) { instance_spy('ActionMailer::Delivery') }
```
You're right, I should have double-checked that. It was from another existing context that I pulled out of this spec below,
```
context 'when disable_builtin_emails is true' do
      it 'does not send the password reset email' do
        allow(SolidusTracking.configuration).to receive(:disable_builtin_emails).and_return(true)
        email = instance_spy('ActionMailer::Delivery')
        allow(Devise.mailer).to receive(:reset_password_instructions).and_return(email)
...
```
That context did expectations against it though, but I am not, so it isn't necessary anymore.

Commit:
Remove unneeded decorator

No longer required with Solidus 4+. The method it is overriding doesn't
exist anymore.

Commit:
Fix various failing specs

- Add the requirement of open struct now that it isn't automatically
  included
- Change expectations to account for changes with how state changes are
  handled in background jobs. solidusio/solidus@2a2df71
- Replace build_stubbed with just build, some objects were missing
  fields
- Since we removed the requirement for solidus_frontend from this gem,
  we don't care about the reset password email sending so we can just
  stub it out

Commit:
Add decorator for DisableCancelEmail

Similar to the confirm email, we need this to respect the
'disable_builtin_emails' value.

jarednorman commented:
I've just disabled the CircleCI builds, but it looks like the GitHub builds are busted.

_Edit: I see you are working on that in #42._

adammathys committed:
Update Github actions setup

Now that we're only supporting Solidus v4.0-4.5 with these changes, we
need to tweak our testing matrix.

Also, we were missing some configuration in order for our Github action
environment variables to be able to pull the correct versions of Rails
and Solidus.

adammathys committed:
Add run migrations step to install generator

At some point in Rails 7.1+, the dummy app generation changed such that
it no longer would fully setup the database for us. All of our other
extensions seem to have avoided this issue because they all contain some
migrations and run them as part of their install steps.

So, even though we don't currently add any migrations from this
extension, we're adding the step so that our CI can correctly setup the
test database.

adammathys committed:
Bump to v1.0.0

We're cutting off support for older versions of Solidus so we should do
a major version bump. (Even if it appears we have never changed the
version of this extension since inception.)

adammathys committed:
Update README with Github actions badge

We're no longer using CircleCI for tests, so we should switch the badge
to something more useful.

forkata reviewed (approved):
Great work on these changes @kjriga and @adammathys 🎉
